pkg/trace/stats/client_stats_aggregator.go

```diff
@@ -0,0 +1,283 @@
+func NewClientStatsAggregator(conf *config.AgentConfig, out chan pb.StatsPayload) *ClientStatsAggregator {
+	oldestTSUnix := uint64(time.Now().Add(-10 * time.Second).Unix())
+	return &ClientStatsAggregator{
+		flushTicker:   time.NewTicker(time.Second),
+		In:            make(chan pb.ClientStatsPayload, 10),
+		buckets:       make(map[uint64]*bucket, 20),
+		out:           out,
+		agentEnv:      conf.DefaultEnv,
+		agentHostname: conf.Hostname,
+		oldestTs:      oldestTSUnix,
+	}
+}
+
+// Start starts the aggregator.
+func (a *ClientStatsAggregator) Start() {
+	go func() {
+		defer watchdog.LogOnPanic()
+		a.run()
+	}()
+}
+
+// Stop stops the aggregator.
+func (a *ClientStatsAggregator) Stop() {
+	close(a.exit)
+	<-a.done
+}
+
+func (a *ClientStatsAggregator) run() {
+	for {
+		select {
+		case t := <-a.flushTicker.C:
+			a.flushOnTime(t)
+		case input := <-a.In:
+			a.add(time.Now(), input)
+		case <-a.exit:
+			a.flushAll()
+			close(a.done)
+		}
+	}
+}
+
+func (a *ClientStatsAggregator) flushOnTime(t time.Time) {
+	flushTs := uint64(t.Unix()) - oldestBucketStartSecond
+	for ts := a.oldestTs; ts < flushTs; ts += 1 {
+		if b, ok := a.buckets[ts]; ok {
+			for _, p := range b.flush() {
+				a.flush(p)
+			}
+			delete(a.buckets, ts)
+		}
+	}
+	a.oldestTs = flushTs
+}
+
+func (a *ClientStatsAggregator) flushAll() {
+	for _, b := range a.buckets {
+		for _, p := range b.flush() {
+			a.flush(p)
+		}
+	}
+}
+
+// getAggregationBucketTime returns unix time at which we aggregate the bucket.
+// We timeshift payloads older than a.oldestTs to a.oldestTs.
+// Payloads in the future are timeshifted to time.Now().Unix()
+func (a *ClientStatsAggregator) getAggregationBucketTime(now time.Time, bucketStart uint64) (uint64, bool) {
+	tsUnix := bucketStart / aggBucketSize
+	if tsUnix < a.oldestTs {
+		return a.oldestTs, true
+	}
+	nowUnix := uint64(now.Unix())
+	if tsUnix > nowUnix {
+		return nowUnix, true
+	}
+	return bucketStart / aggBucketSize, false
+}
+
+func (a *ClientStatsAggregator) add(now time.Time, p pb.ClientStatsPayload) {
+	clientStats := p.Stats
+	for _, clientBucket := range clientStats {
+		tsUnix, timeShifted := a.getAggregationBucketTime(now, clientBucket.Start)
+		if timeShifted {
+			newStart := tsUnix * 1e9
+			clientBucket.AgentTimeShift = int64(tsUnix) - int64(clientBucket.Start)
+			clientBucket.Start = newStart
+		}
+		b, ok := a.buckets[tsUnix]
+		if !ok {
+			b = &bucket{tsUnix: tsUnix}
+			a.buckets[tsUnix] = b
+		}
+		p.Stats = []pb.ClientStatsBucket{clientBucket}
+		for _, p := range b.add(p) {
+			a.flush(p)
+		}
+	}
+}
+
+func (a *ClientStatsAggregator) flush(p pb.ClientStatsPayload) {
+	a.out <- pb.StatsPayload{
+		Stats:          []pb.ClientStatsPayload{p},
+		AgentEnv:       a.agentEnv,
+		AgentHostname:  a.agentHostname,
+		AgentVersion:   info.Version,
+		ClientComputed: true,
+	}
+}
+
+type bucket struct {
+	firstPayload pb.ClientStatsPayload
+	tsUnix       uint64
```

Comment:
```suggestion
	ts       uint64
```
Gotta ask again - do we need the Unix suffix everywhere?